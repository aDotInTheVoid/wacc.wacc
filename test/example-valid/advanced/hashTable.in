d
a
1
a
2
a
3
d
e
3
d
f
d
g