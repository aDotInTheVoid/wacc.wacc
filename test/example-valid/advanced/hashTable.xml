<unit>
  <fn>
    <name>init</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <local>0</local>
      <len/>
      <assign_do/>
      <addr_local>2</addr_local>
      <num>0</num>
      <assign_do/>
      <while>
        <cond><local>2</local><local>1</local><op>lt</op></cond>
        <body>
          <addr_local>0</addr_local>
          <local>2</local>
          <addr_array/>
          <null/>
          <assign_do/>
          <addr_local>2</addr_local>
          <local>2</local>
          <num>1</num>
          <op>add</op>
          <assign_do/>
        </body>
      </while>
      <true/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>contain</name>
    <args><arg>0</arg><arg>1</arg></args>
    <body>
      <addr_local>2</addr_local>
      <local>0</local>
      <local>1</local>
      <call_func><name>calculateIndex</name><nargs>2</nargs></call_func>
      <assign_do/>
      <addr_local>3</addr_local>
      <local>0</local>
      <local>2</local>
      <array_elem/>
      <local>1</local>
      <call_func><name>findNode</name><nargs>2</nargs></call_func>
      <assign_do/>
      <local>3</local>
      <null/>
      <op>ne</op>
      <return/>
    </body>
  </fn>
  <fn>
    <name>insertIfNotContain</name>
    <args><arg>0</arg><arg>1</arg></args>
    <body>
      <addr_local>2</addr_local>
      <local>0</local>
      <local>1</local>
      <call_func><name>calculateIndex</name><nargs>2</nargs></call_func>
      <assign_do/>
      <addr_local>3</addr_local>
      <local>0</local>
      <local>2</local>
      <array_elem/>
      <local>1</local>
      <call_func><name>findNode</name><nargs>2</nargs></call_func>
      <assign_do/>
      <if>
        <cond><local>3</local><null/><op>ne</op></cond>
        <when><false/><return/></when>
        <else>
          <addr_local>4</addr_local>
          <local>1</local>
          <local>0</local>
          <local>2</local>
          <array_elem/>
          <newpair/>
          <assign_do/>
          <addr_local>0</addr_local>
          <local>2</local>
          <addr_array/>
          <local>4</local>
          <assign_do/>
          <true/>
          <return/>
        </else>
      </if>
    </body>
  </fn>
  <fn>
    <name>remove</name>
    <args><arg>0</arg><arg>1</arg></args>
    <body>
      <addr_local>2</addr_local>
      <local>0</local>
      <local>1</local>
      <call_func><name>calculateIndex</name><nargs>2</nargs></call_func>
      <assign_do/>
      <addr_local>3</addr_local>
      <local>0</local>
      <local>2</local>
      <array_elem/>
      <local>1</local>
      <call_func><name>findNode</name><nargs>2</nargs></call_func>
      <assign_do/>
      <if>
        <cond><local>3</local><null/><op>eq</op></cond>
        <when><false/><return/></when>
        <else>
          <addr_local>0</addr_local>
          <local>2</local>
          <addr_array/>
          <local>0</local>
          <local>2</local>
          <array_elem/>
          <local>3</local>
          <call_func><name>removeNode</name><nargs>2</nargs></call_func>
          <assign_do/>
          <true/>
          <return/>
        </else>
      </if>
    </body>
  </fn>
  <fn>
    <name>removeAll</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <local>0</local>
      <len/>
      <assign_do/>
      <addr_local>2</addr_local>
      <num>0</num>
      <assign_do/>
      <while>
        <cond><local>2</local><local>1</local><op>lt</op></cond>
        <body>
          <addr_local>3</addr_local>
          <local>0</local>
          <local>2</local>
          <array_elem/>
          <assign_do/>
          <while>
            <cond><local>3</local><null/><op>ne</op></cond>
            <body>
              <addr_local>4</addr_local>
              <local>3</local>
              <snd/>
              <assign_do/>
              <local>3</local>
              <free/>
              <addr_local>3</addr_local>
              <local>4</local>
              <assign_do/>
            </body>
          </while>
          <addr_local>0</addr_local>
          <local>2</local>
          <addr_array/>
          <null/>
          <assign_do/>
          <addr_local>2</addr_local>
          <local>2</local>
          <num>1</num>
          <op>add</op>
          <assign_do/>
        </body>
      </while>
      <true/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>count</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <local>0</local>
      <len/>
      <assign_do/>
      <addr_local>2</addr_local>
      <num>0</num>
      <assign_do/>
      <addr_local>3</addr_local>
      <num>0</num>
      <assign_do/>
      <while>
        <cond><local>3</local><local>1</local><op>lt</op></cond>
        <body>
          <addr_local>4</addr_local>
          <local>0</local>
          <local>3</local>
          <array_elem/>
          <call_func><name>countNodes</name><nargs>1</nargs></call_func>
          <assign_do/>
          <addr_local>2</addr_local>
          <local>2</local>
          <local>4</local>
          <op>add</op>
          <assign_do/>
          <addr_local>3</addr_local>
          <local>3</local>
          <num>1</num>
          <op>add</op>
          <assign_do/>
        </body>
      </while>
      <local>2</local>
      <return/>
    </body>
  </fn>
  <fn>
    <name>printAll</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <local>0</local>
      <len/>
      <assign_do/>
      <addr_local>2</addr_local>
      <num>0</num>
      <assign_do/>
      <while>
        <cond><local>2</local><local>1</local><op>lt</op></cond>
        <body>
          <addr_local>3</addr_local>
          <local>0</local>
          <local>2</local>
          <array_elem/>
          <call_func><name>printAllNodes</name><nargs>1</nargs></call_func>
          <assign_do/>
          <addr_local>2</addr_local>
          <local>2</local>
          <num>1</num>
          <op>add</op>
          <assign_do/>
        </body>
      </while>
      <str>""</str>
      <print><kind>str</kind><newline>true</newline></print>
      <true/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>calculateIndex</name>
    <args><arg>0</arg><arg>1</arg></args>
    <body>
      <addr_local>2</addr_local>
      <local>0</local>
      <len/>
      <assign_do/>
      <local>1</local>
      <local>2</local>
      <op>mod</op>
      <return/>
    </body>
  </fn>
  <fn>
    <name>findNode</name>
    <args><arg>0</arg><arg>1</arg></args>
    <body>
      <while>
        <cond><local>0</local><null/><op>ne</op></cond>
        <body>
          <addr_local>2</addr_local>
          <local>0</local>
          <fst/>
          <assign_do/>
          <if>
            <cond><local>2</local><local>1</local><op>eq</op></cond>
            <when><local>0</local><return/></when>
            <else><addr_local>0</addr_local><local>0</local><snd/><assign_do/></else>
          </if>
        </body>
      </while>
      <null/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>removeNode</name>
    <args><arg>0</arg><arg>1</arg></args>
    <body>
      <if>
        <cond><local>0</local><null/><op>eq</op></cond>
        <when><null/><return/></when>
        <else>
          <if>
            <cond><local>0</local><local>1</local><op>eq</op></cond>
            <when>
              <addr_local>0</addr_local>
              <local>0</local>
              <snd/>
              <assign_do/>
              <local>1</local>
              <free/>
              <local>0</local>
              <return/>
            </when>
            <else>
              <addr_local>2</addr_local>
              <local>0</local>
              <snd/>
              <assign_do/>
              <addr_local>0</addr_local>
              <addr_snd/>
              <local>2</local>
              <local>1</local>
              <call_func><name>removeNode</name><nargs>2</nargs></call_func>
              <assign_do/>
              <local>0</local>
              <return/>
            </else>
          </if>
        </else>
      </if>
    </body>
  </fn>
  <fn>
    <name>countNodes</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <num>0</num>
      <assign_do/>
      <while>
        <cond><local>0</local><null/><op>ne</op></cond>
        <body>
          <addr_local>1</addr_local>
          <local>1</local>
          <num>1</num>
          <op>add</op>
          <assign_do/>
          <addr_local>0</addr_local>
          <local>0</local>
          <snd/>
          <assign_do/>
        </body>
      </while>
      <local>1</local>
      <return/>
    </body>
  </fn>
  <fn>
    <name>printAllNodes</name>
    <args><arg>0</arg></args>
    <body>
      <while>
        <cond><local>0</local><null/><op>ne</op></cond>
        <body>
          <addr_local>1</addr_local>
          <local>0</local>
          <fst/>
          <assign_do/>
          <local>1</local>
          <print><kind>i32</kind><newline>false</newline></print>
          <char>' '</char>
          <print><kind>char</kind><newline>false</newline></print>
          <addr_local>0</addr_local>
          <local>0</local>
          <snd/>
          <assign_do/>
        </body>
      </while>
      <true/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>printMenu</name>
    <args/>
    <body>
      <str>"==========================================="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"========== Hash Table Program ============="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"==========================================="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"=                                         ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"= Please choose the following options:    ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"=                                         ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"= a: insert an integer                    ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"= b: find an integer                      ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"= c: count the integers                   ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"= d: print all integers                   ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"= e: remove an integer                    ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"= f: remove all integers                  ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"= g: exit                                 ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"=                                         ="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <str>"==========================================="</str>
      <print><kind>str</kind><newline>true</newline></print>
      <addr_local>0</addr_local>
      <char>'a'</char>
      <assign_do/>
      <addr_local>1</addr_local>
      <char>'g'</char>
      <assign_do/>
      <while>
        <cond><true/></cond>
        <body>
          <str>"Your decision: "</str>
          <print><kind>str</kind><newline>false</newline></print>
          <addr_local>2</addr_local>
          <char>'\0'</char>
          <assign_do/>
          <addr_local>2</addr_local>
          <read><ty>char</ty></read>
          <addr_local>3</addr_local>
          <local>2</local>
          <assign_do/>
          <if>
            <cond>
              <local>0</local>
              <local>3</local>
              <op>le</op>
              <local>3</local>
              <local>1</local>
              <op>le</op>
              <op>and</op>
            </cond>
            <when><local>2</local><return/></when>
            <else>
              <str>"You have entered: "</str>
              <print><kind>str</kind><newline>false</newline></print>
              <local>2</local>
              <print><kind>char</kind><newline>false</newline></print>
              <str>" which is invalid, please try again."</str>
              <print><kind>str</kind><newline>true</newline></print>
            </else>
          </if>
        </body>
      </while>
      <char>'\0'</char>
      <return/>
    </body>
  </fn>
  <fn>
    <name>askForInt</name>
    <args><arg>0</arg></args>
    <body>
      <local>0</local>
      <print><kind>str</kind><newline>false</newline></print>
      <addr_local>1</addr_local>
      <num>0</num>
      <assign_do/>
      <addr_local>1</addr_local>
      <read><ty>int</ty></read>
      <str>"You have entered: "</str>
      <print><kind>str</kind><newline>false</newline></print>
      <local>1</local>
      <print><kind>i32</kind><newline>true</newline></print>
      <local>1</local>
      <return/>
    </body>
  </fn>
  <fn>
    <name>handleMenuInsert</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <str>"Please enter an integer to insert: "</str>
      <call_func><name>askForInt</name><nargs>1</nargs></call_func>
      <assign_do/>
      <addr_local>2</addr_local>
      <local>0</local>
      <local>1</local>
      <call_func><name>insertIfNotContain</name><nargs>2</nargs></call_func>
      <assign_do/>
      <if>
        <cond><local>2</local></cond>
        <when>
          <str>"Successfully insert it. The integer is new."</str>
          <print><kind>str</kind><newline>true</newline></print>
        </when>
        <else>
          <str>"The integer is already there. No insertion is made."</str>
          <print><kind>str</kind><newline>true</newline></print>
        </else>
      </if>
      <true/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>handleMenuFind</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <str>"Please enter an integer to find: "</str>
      <call_func><name>askForInt</name><nargs>1</nargs></call_func>
      <assign_do/>
      <addr_local>2</addr_local>
      <local>0</local>
      <local>1</local>
      <call_func><name>contain</name><nargs>2</nargs></call_func>
      <assign_do/>
      <if>
        <cond><local>2</local></cond>
        <when>
          <str>"Find the integer."</str>
          <print><kind>str</kind><newline>true</newline></print>
        </when>
        <else>
          <str>"The integer is not found."</str>
          <print><kind>str</kind><newline>true</newline></print>
        </else>
      </if>
      <true/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>handleMenuCount</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <local>0</local>
      <call_func><name>count</name><nargs>1</nargs></call_func>
      <assign_do/>
      <if>
        <cond><local>1</local><num>1</num><op>eq</op></cond>
        <when>
          <str>"There is only 1 integer."</str>
          <print><kind>str</kind><newline>true</newline></print>
        </when>
        <else>
          <str>"There are "</str>
          <print><kind>str</kind><newline>false</newline></print>
          <local>1</local>
          <print><kind>i32</kind><newline>false</newline></print>
          <str>" integers."</str>
          <print><kind>str</kind><newline>true</newline></print>
        </else>
      </if>
      <true/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>handleMenuPrint</name>
    <args><arg>0</arg></args>
    <body>
      <str>"Here are the integers: "</str>
      <print><kind>str</kind><newline>false</newline></print>
      <addr_local>1</addr_local>
      <local>0</local>
      <call_func><name>printAll</name><nargs>1</nargs></call_func>
      <assign_do/>
      <true/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>handleMenuRemove</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <str>"Please enter an integer to remove: "</str>
      <call_func><name>askForInt</name><nargs>1</nargs></call_func>
      <assign_do/>
      <addr_local>2</addr_local>
      <local>0</local>
      <local>1</local>
      <call_func><name>remove</name><nargs>2</nargs></call_func>
      <assign_do/>
      <if>
        <cond><local>2</local></cond>
        <when>
          <str>"The integer has been removed."</str>
          <print><kind>str</kind><newline>true</newline></print>
        </when>
        <else>
          <str>"The integer is not found."</str>
          <print><kind>str</kind><newline>true</newline></print>
        </else>
      </if>
      <true/>
      <return/>
    </body>
  </fn>
  <fn>
    <name>handleMenuRemoveAll</name>
    <args><arg>0</arg></args>
    <body>
      <addr_local>1</addr_local>
      <local>0</local>
      <call_func><name>removeAll</name><nargs>1</nargs></call_func>
      <assign_do/>
      <str>"All integers have been removed."</str>
      <print><kind>str</kind><newline>true</newline></print>
      <true/>
      <return/>
    </body>
  </fn>
  <main>
    <addr_local>0</addr_local>
    <null/>
    <null/>
    <null/>
    <null/>
    <null/>
    <null/>
    <null/>
    <null/>
    <null/>
    <null/>
    <null/>
    <null/>
    <null/>
    <array_lit>13</array_lit>
    <assign_do/>
    <addr_local>1</addr_local>
    <local>0</local>
    <call_func><name>init</name><nargs>1</nargs></call_func>
    <assign_do/>
    <addr_local>2</addr_local>
    <true/>
    <assign_do/>
    <while>
      <cond><local>2</local></cond>
      <body>
        <addr_local>3</addr_local>
        <call_func><name>printMenu</name><nargs>0</nargs></call_func>
        <assign_do/>
        <if>
          <cond><local>3</local><char>'a'</char><op>eq</op></cond>
          <when>
            <addr_local>4</addr_local>
            <local>0</local>
            <call_func><name>handleMenuInsert</name><nargs>1</nargs></call_func>
            <assign_do/>
          </when>
          <else>
            <if>
              <cond><local>3</local><char>'b'</char><op>eq</op></cond>
              <when>
                <addr_local>4</addr_local>
                <local>0</local>
                <call_func><name>handleMenuFind</name><nargs>1</nargs></call_func>
                <assign_do/>
              </when>
              <else>
                <if>
                  <cond><local>3</local><char>'c'</char><op>eq</op></cond>
                  <when>
                    <addr_local>4</addr_local>
                    <local>0</local>
                    <call_func><name>handleMenuCount</name><nargs>1</nargs></call_func>
                    <assign_do/>
                  </when>
                  <else>
                    <if>
                      <cond><local>3</local><char>'d'</char><op>eq</op></cond>
                      <when>
                        <addr_local>4</addr_local>
                        <local>0</local>
                        <call_func><name>handleMenuPrint</name><nargs>1</nargs></call_func>
                        <assign_do/>
                      </when>
                      <else>
                        <if>
                          <cond><local>3</local><char>'e'</char><op>eq</op></cond>
                          <when>
                            <addr_local>4</addr_local>
                            <local>0</local>
                            <call_func><name>handleMenuRemove</name><nargs>1</nargs></call_func>
                            <assign_do/>
                          </when>
                          <else>
                            <if>
                              <cond><local>3</local><char>'f'</char><op>eq</op></cond>
                              <when>
                                <addr_local>4</addr_local>
                                <local>0</local>
                                <call_func><name>handleMenuRemoveAll</name><nargs>1</nargs></call_func>
                                <assign_do/>
                              </when>
                              <else>
                                <if>
                                  <cond><local>3</local><char>'g'</char><op>eq</op></cond>
                                  <when>
                                    <str>"Goodbye Human"</str>
                                    <print><kind>str</kind><newline>true</newline></print>
                                    <addr_local>2</addr_local>
                                    <false/>
                                    <assign_do/>
                                  </when>
                                  <else>
                                    <str>"Error: unknown choice ("</str>
                                    <print><kind>str</kind><newline>false</newline></print>
                                    <local>3</local>
                                    <print><kind>char</kind><newline>false</newline></print>
                                    <str>")"</str>
                                    <print><kind>str</kind><newline>true</newline></print>
                                    <num>1</num>
                                    <neg/>
                                    <exit/>
                                  </else>
                                </if>
                              </else>
                            </if>
                          </else>
                        </if>
                      </else>
                    </if>
                  </else>
                </if>
              </else>
            </if>
          </else>
        </if>
      </body>
    </while>
  </main>
</unit>