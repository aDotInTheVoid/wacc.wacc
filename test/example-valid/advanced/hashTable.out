===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: Here are the integers: 
===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: Please enter an integer to insert: You have entered: 1
Successfully insert it. The integer is new.
===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: Please enter an integer to insert: You have entered: 2
Successfully insert it. The integer is new.
===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: Please enter an integer to insert: You have entered: 3
Successfully insert it. The integer is new.
===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: Here are the integers: 1 2 3 
===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: Please enter an integer to remove: You have entered: 3
The integer has been removed.
===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: Here are the integers: 1 2 
===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: All integers have been removed.
===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: Here are the integers: 
===========================================
========== Hash Table Program =============
===========================================
=                                         =
= Please choose the following options:    =
=                                         =
= a: insert an integer                    =
= b: find an integer                      =
= c: count the integers                   =
= d: print all integers                   =
= e: remove an integer                    =
= f: remove all integers                  =
= g: exit                                 =
=                                         =
===========================================
Your decision: Goodbye Human
